File: class19/java/PES1904-CleanOnionHexagonal.java
/*
 * PES1903 - Clean, Onion e Hexagonal: Comparativo com o Mesmo Domínio
 * Aula 19: Camadas, MVC, Clean Architecture e Hexagonal
 * Projeto e Engenharia de Software - Prof. Giovane Barcelos
 *
 * Exemplo comparativo: o mesmo domínio (Pedido) estruturado em três
 * estilos arquiteturais — Clean, Onion e Hexagonal — mostrando como
 * cada um organiza as dependências e camadas.
 *
 * Domínio comum: Gerenciamento de pedidos com ItemPedido e cálculo de total.
 *
 * Compilar/executar:
 *   javac PES1903-CleanOnionHexagonal.java && java CleanOnionHexagonal
 */

import java.util.*;

// ============================================================
// DOMÍNIO COMPARTILHADO (entidades, value objects, regras)
// ============================================================

class Dinheiro {
    final double valor;

    Dinheiro(double valor) { this.valor = valor; }

    Dinheiro somar(Dinheiro outro) { return new Dinheiro(this.valor + outro.valor); }

    @Override
    public String toString() { return String.format("R$ %,.2f", valor); }
}

class ItemPedido {
    final String produtoId, nome;
    final Dinheiro preco;
    final int quantidade;

    ItemPedido(String produtoId, String nome, Dinheiro preco, int quantidade) {
        this.produtoId = produtoId;
        this.nome = nome;
        this.preco = preco;
        this.quantidade = quantidade;
    }

    Dinheiro subtotal() { return new Dinheiro(preco.valor * quantidade); }

    @Override
    public String toString() {
        return String.format("%s x%d %s", nome, quantidade, subtotal());
    }
}

class Pedido {
    final int pedidoId;
    final String cliente;
    private final List<ItemPedido> itens = new ArrayList<>();
    private String status = "ABERTO";

    Pedido(int pedidoId, String cliente) {
        this.pedidoId = pedidoId;
        this.cliente = cliente;
    }

    void adicionarItem(String produtoId, String nome, double preco, int qtd) {
        itens.add(new ItemPedido(produtoId, nome, new Dinheiro(preco), qtd));
    }

    Dinheiro total() {
        Dinheiro t = new Dinheiro(0);
        for (ItemPedido i : itens) t = t.somar(i.subtotal());
        return t;
    }

    void fechar() {
        if (itens.isEmpty()) throw new IllegalStateException("Pedido sem itens.");
        status = "FECHADO";
    }

    String getStatus() { return status; }
    List<ItemPedido> getItens() { return Collections.unmodifiableList(itens); }

    @Override
    public String toString() {
        return String.format("Pedido[%d] %s — %s — %s", pedidoId, cliente, status, total());
    }
}

// ============================================================
// INTERFACE DE REPOSITÓRIO (porta — usada por todos os estilos)
// ============================================================

interface PedidoRepository {
    void salvar(Pedido pedido);
    Pedido buscarPorId(int pedidoId);
}

// ============================================================
// 1. CLEAN ARCHITECTURE
// ============================================================

class PedidoRepositoryEmMemoria implements PedidoRepository {
    private final Map<Integer, Pedido> db = new HashMap<>();

    public void salvar(Pedido p) { db.put(p.pedidoId, p); }
    public Pedido buscarPorId(int id) { return db.get(id); }
}

// Caso de Uso (Application Layer)
class CriarPedidoUseCase {
    private final PedidoRepository repo;

    CriarPedidoUseCase(PedidoRepository repo) { this.repo = repo; }

    Pedido executar(int id, String cliente, Object[][] itens) {
        Pedido p = new Pedido(id, cliente);
        for (Object[] item : itens) {
            p.adicionarItem((String) item[0], (String) item[1],
                (Double) item[2], (Integer) item[3]);
        }
        p.fechar();
        repo.salvar(p);
        return p;
    }
}

// Controller (Interface Adapter)
class PedidoController {
    private final CriarPedidoUseCase usecase;

    PedidoController(CriarPedidoUseCase usecase) { this.usecase = usecase; }

    String criarPedido(int id, String cliente, Object[][] itens) {
        Pedido p = usecase.executar(id, cliente, itens);
        return String.format("[Clean] Pedido #%d | %s | %s | %s",
            p.pedidoId, p.cliente, p.total(), p.getStatus());
    }
}

// ============================================================
// 2. ONION ARCHITECTURE
// ============================================================

class OnionPedidoService {
    private final PedidoRepository repo;

    OnionPedidoService(PedidoRepository repo) { this.repo = repo; }

    Pedido criarEFechar(int id, String cliente, Object[][] itens) {
        Pedido p = new Pedido(id, cliente);
        for (Object[] item : itens) {
            p.adicionarItem((String) item[0], (String) item[1],
                (Double) item[2], (Integer) item[3]);
        }
        p.fechar();
        repo.salvar(p);
        return p;
    }

    String resumo(int id) {
        Pedido p = repo.buscarPorId(id);
        if (p == null) return null;
        return String.format("[Onion] Pedido #%d — %s — %s (%s)",
            p.pedidoId, p.cliente, p.total(), p.getStatus());
    }
}

// ============================================================
// 3. HEXAGONAL ARCHITECTURE (Ports & Adapters)
// ============================================================

class HexagonalPedidoHandler {
    private final PedidoRepository repo;

    HexagonalPedidoHandler(PedidoRepository repo) { this.repo = repo; }

    String handleCriarPedido(Map<String, Object> command) {
        int id = (Integer) command.get("pedido_id");
        String cliente = (String) command.get("cliente");
        @SuppressWarnings("unchecked")
        List<Map<String, Object>> itens = (List<Map<String, Object>>) command.get("itens");

        Pedido p = new Pedido(id, cliente);
        for (Map<String, Object> item : itens) {
            p.adicionarItem((String) item.get("id"), (String) item.get("nome"),
                (Double) item.get("preco"), (Integer) item.get("qtd"));
        }
        p.fechar();
        repo.salvar(p);

        return String.format("[Hexagonal] Pedido #%d | %s | %s | %s",
            p.pedidoId, p.cliente, p.total(), p.getStatus());
    }

    String handleConsultar(int id) {
        Pedido p = repo.buscarPorId(id);
        if (p == null) return "Pedido não encontrado.";
        StringBuilder sb = new StringBuilder();
        sb.append(String.format("[Hexagonal] #%d — %s — %s (%s) | Itens: ",
            p.pedidoId, p.cliente, p.total(), p.getStatus()));
        for (ItemPedido i : p.getItens())
            sb.append(i.nome).append(" x").append(i.quantidade).append(", ");
        return sb.substring(0, sb.length() - 2);
    }
}

// ============================================================
// DEMONSTRAÇÃO COMPARATIVA
// ============================================================

class CleanOnionHexagonal {

    static void demoClean() {
        System.out.println("\n" + "─".repeat(55));
        System.out.println("  1. CLEAN ARCHITECTURE");
        System.out.println("     Entities → Use Cases → Controllers → Repositories");
        System.out.println("─".repeat(55));

        PedidoRepository repo = new PedidoRepositoryEmMemoria();
        CriarPedidoUseCase usecase = new CriarPedidoUseCase(repo);
        PedidoController controller = new PedidoController(usecase);

        Object[][] itens = {
            {"P01", "Monitor 24\"", 1200.00, 1},
            {"P02", "Teclado Mecânico", 350.00, 2},
        };
        System.out.println("  " + controller.criarPedido(1001, "Carlos Oliveira", itens));
        System.out.println("  ✓ Domínio isolado → Controller → UseCase → Repository");
    }

    static void demoOnion() {
        System.out.println("\n" + "─".repeat(55));
        System.out.println("  2. ONION ARCHITECTURE");
        System.out.println("     Domain ← Application ← Infrastructure ← Presentation");
        System.out.println("─".repeat(55));

        PedidoRepository repo = new PedidoRepositoryEmMemoria();
        OnionPedidoService service = new OnionPedidoService(repo);

        Object[][] itens = {
            {"P03", "Notebook", 4500.00, 1},
            {"P04", "Mouse", 120.00, 1},
        };
        service.criarEFechar(2001, "Ana Beatriz", itens);
        System.out.println("  " + service.resumo(2001));
        System.out.println("  ✓ Camadas concêntricas: dependência sempre para dentro");
    }

    static void demoHexagonal() {
        System.out.println("\n" + "─".repeat(55));
        System.out.println("  3. HEXAGONAL (PORTS & ADAPTERS)");
        System.out.println("     Portas (interfaces) + Adaptadores (concretos)");
        System.out.println("─".repeat(55));

        PedidoRepository repo = new PedidoRepositoryEmMemoria();
        HexagonalPedidoHandler handler = new HexagonalPedidoHandler(repo);

        Map<String, Object> cmd = new HashMap<>();
        cmd.put("pedido_id", 3001);
        cmd.put("cliente", "Roberto Lima");
        List<Map<String, Object>> itens = new ArrayList<>();
        itens.add(Map.of("id", "P05", "nome", "SSD 1TB", "preco", 600.00, "qtd", 2));
        itens.add(Map.of("id", "P06", "nome", "Cabo HDMI", "preco", 45.00, "qtd", 3));
        cmd.put("itens", itens);

        System.out.println("  " + handler.handleCriarPedido(cmd));
        System.out.println("  " + handler.handleConsultar(3001));
        System.out.println("  ✓ Domínio 100% isolado por portas; adaptadores trocáveis");
    }

    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("  CLEAN × ONION × HEXAGONAL — MESMO DOMÍNIO");
        System.out.println("=".repeat(60));
        System.out.println("\n  Domínio comum: Pedido + ItemPedido + Dinheiro (Value Object)");

        demoClean();
        demoOnion();
        demoHexagonal();

        System.out.println("\n" + "=".repeat(60));
        System.out.println("  COMPARATIVO:");
        System.out.println("  ┌──────────┬──────────────────────────────────┐");
        System.out.println("  │ Clean    │ Entities → UseCases → Adapters   │");
        System.out.println("  │ Onion    │ Domain ← App ← Infra ← Present.  │");
        System.out.println("  │ Hexagonal│ Ports (interfaces) + Adapters     │");
        System.out.println("  ├──────────┼──────────────────────────────────┤");
        System.out.println("  │ TODOS    │ Domínio isolado de frameworks     │");
        System.out.println("  │ TODOS    │ Dependência aponta para dentro    │");
        System.out.println("  │ TODOS    │ Testabilidade e troca de infra    │");
        System.out.println("  └──────────┴──────────────────────────────────┘");
        System.out.println("=".repeat(60));
    }
}

File: class19/python/PES1904-CleanOnionHexagonal.py
"""
PES1903 - Clean, Onion e Hexagonal: Comparativo com o Mesmo Domínio
Aula 19: Camadas, MVC, Clean Architecture e Hexagonal
Projeto e Engenharia de Software - Prof. Giovane Barcelos

Exemplo comparativo: o mesmo domínio (Pedido) estruturado em três
estilos arquiteturais — Clean, Onion e Hexagonal — mostrando como
cada um organiza as dependências e camadas.

Domínio comum: Gerenciamento de pedidos com ItemPedido e cálculo de total.
"""

from __future__ import annotations
from abc import ABC, abstractmethod
from dataclasses import dataclass
from typing import List


# ============================================================
# DOMÍNIO COMPARTILHADO (entidades, value objects, regras)
# ============================================================

@dataclass(frozen=True)
class Dinheiro:
    """Value Object imutável."""
    valor: float

    def somar(self, outro: "Dinheiro") -> "Dinheiro":
        return Dinheiro(self.valor + outro.valor)

    def __repr__(self) -> str:
        return f"R$ {self.valor:,.2f}"


@dataclass(frozen=True)
class ItemPedido:
    """Entity — com identidade de produto."""
    produto_id: str
    nome: str
    preco: Dinheiro
    quantidade: int

    def subtotal(self) -> Dinheiro:
        return Dinheiro(self.preco.valor * self.quantidade)


class Pedido:
    """Aggregate Root — entidade raiz do agregado."""

    def __init__(self, pedido_id: int, cliente: str):
        self.pedido_id = pedido_id
        self.cliente = cliente
        self._itens: List[ItemPedido] = []
        self._status = "ABERTO"

    def adicionar_item(self, produto_id: str, nome: str, preco: float, quantidade: int) -> None:
        self._itens.append(ItemPedido(produto_id, nome, Dinheiro(preco), quantidade))

    def total(self) -> Dinheiro:
        t = Dinheiro(0.0)
        for item in self._itens:
            t = t.somar(item.subtotal())
        return t

    @property
    def status(self) -> str:
        return self._status

    @property
    def itens(self) -> List[ItemPedido]:
        return list(self._itens)

    def fechar(self) -> None:
        if not self._itens:
            raise ValueError("Pedido sem itens não pode ser fechado.")
        self._status = "FECHADO"


# ============================================================
# INTERFACE DE REPOSITÓRIO (porta — usada por todos os estilos)
# ============================================================

class PedidoRepository(ABC):
    """Porta: abstração de persistência (usada em todos os estilos)."""

    @abstractmethod
    def salvar(self, pedido: Pedido) -> None:
        ...

    @abstractmethod
    def buscar_por_id(self, pedido_id: int) -> Pedido | None:
        ...


# ============================================================
# 1. CLEAN ARCHITECTURE
#    Estrutura: Entities → Use Cases → Interface Adapters → Frameworks
# ============================================================

# -- Caso de Uso (Application Layer) --
class CriarPedidoUseCase:
    """Caso de uso: orquestra a criação de pedido."""

    def __init__(self, repo: PedidoRepository):
        self._repo = repo

    def executar(self, pedido_id: int, cliente: str, itens: List[tuple]) -> Pedido:
        pedido = Pedido(pedido_id, cliente)
        for prod_id, nome, preco, qtd in itens:
            pedido.adicionar_item(prod_id, nome, preco, qtd)
        pedido.fechar()
        self._repo.salvar(pedido)
        return pedido


# -- Adaptador de Interface (Controller) --
class PedidoController:
    """Controller: adapta entrada externa para o caso de uso."""

    def __init__(self, usecase: CriarPedidoUseCase):
        self._usecase = usecase

    def criar_pedido(self, pedido_id: int, cliente: str, itens: List[dict]) -> dict:
        tuplas = [(i["id"], i["nome"], i["preco"], i["qtd"]) for i in itens]
        pedido = self._usecase.executar(pedido_id, cliente, tuplas)
        return {
            "id": pedido.pedido_id,
            "cliente": pedido.cliente,
            "total": str(pedido.total()),
            "status": pedido.status,
        }


# -- Adaptador de Framework (Repository concreto) --
class PedidoRepositoryEmMemoria(PedidoRepository):
    """Implementação concreta (em memória) — camada de infraestrutura."""

    def __init__(self):
        self._db: dict[int, Pedido] = {}

    def salvar(self, pedido: Pedido) -> None:
        self._db[pedido.pedido_id] = pedido

    def buscar_por_id(self, pedido_id: int) -> Pedido | None:
        return self._db.get(pedido_id)


# ============================================================
# 2. ONION ARCHITECTURE
#    Estrutura: Domain → Application → Infrastructure → Presentation
#    Mesma lógica, organização diferente (ênfase em camadas concêntricas)
# ============================================================

class OnionPedidoService:
    """
    Onion Application Service.
    Na Onion, Application depende apenas de Domain (interfaces).
    """

    def __init__(self, repo: PedidoRepository):
        self._repo = repo

    def criar_e_fechar(self, pedido_id: int, cliente: str, itens: List[tuple]) -> Pedido:
        pedido = Pedido(pedido_id, cliente)
        for pid, nome, preco, qtd in itens:
            pedido.adicionar_item(pid, nome, preco, qtd)
        pedido.fechar()
        self._repo.salvar(pedido)
        return pedido

    def resumo(self, pedido_id: int) -> str | None:
        pedido = self._repo.buscar_por_id(pedido_id)
        if pedido is None:
            return None
        return (
            f"[Onion] Pedido #{pedido.pedido_id} — {pedido.cliente} — "
            f"{pedido.total()} ({pedido.status})"
        )


# ============================================================
# 3. HEXAGONAL ARCHITECTURE (Ports & Adapters)
#    Portas (interfaces) já definidas: PedidoRepository
#    Adaptadores concretos conectam o domínio ao mundo externo
# ============================================================

class HexagonalPedidoHandler:
    """
    Handler hexagonal: recebe comando pela porta e processa no domínio.
    O domínio não sabe nada sobre o exterior.
    """

    def __init__(self, repo: PedidoRepository):
        self._repo = repo

    def handle_criar_pedido(self, command: dict) -> dict:
        pedido = Pedido(command["pedido_id"], command["cliente"])
        for item in command["itens"]:
            pedido.adicionar_item(
                item["id"], item["nome"], item["preco"], item["qtd"]
            )
        pedido.fechar()
        self._repo.salvar(pedido)

        # Projeção de resposta (adapter de saída)
        return {
            "pedido_id": pedido.pedido_id,
            "cliente": pedido.cliente,
            "total": str(pedido.total()),
            "status": pedido.status,
        }

    def handle_consultar_pedido(self, pedido_id: int) -> dict | None:
        pedido = self._repo.buscar_por_id(pedido_id)
        if pedido is None:
            return None
        return {
            "pedido_id": pedido.pedido_id,
            "cliente": pedido.cliente,
            "total": str(pedido.total()),
            "status": pedido.status,
            "itens": [f"{i.nome} x{i.quantidade}" for i in pedido.itens],
        }


# ============================================================
# DEMONSTRAÇÃO COMPARATIVA
# ============================================================

def demo_clean():
    print("\n" + "─" * 55)
    print("  1. CLEAN ARCHITECTURE")
    print("     Entities → Use Cases → Controllers → Repositories")
    print("─" * 55)

    repo = PedidoRepositoryEmMemoria()
    usecase = CriarPedidoUseCase(repo)
    controller = PedidoController(usecase)

    result = controller.criar_pedido(1001, "Carlos Oliveira", [
        {"id": "P01", "nome": "Monitor 24'", "preco": 1200.00, "qtd": 1},
        {"id": "P02", "nome": "Teclado Mecânico", "preco": 350.00, "qtd": 2},
    ])
    print(f"  Resultado: {result}")
    print("  ✓ Domínio isolado → Controller → UseCase → Repository")


def demo_onion():
    print("\n" + "─" * 55)
    print("  2. ONION ARCHITECTURE")
    print("     Domain ← Application ← Infrastructure ← Presentation")
    print("─" * 55)

    repo = PedidoRepositoryEmMemoria()
    service = OnionPedidoService(repo)

    service.criar_e_fechar(2001, "Ana Beatriz", [
        ("P03", "Notebook", 4500.00, 1),
        ("P04", "Mouse", 120.00, 1),
    ])
    print(f"  {service.resumo(2001)}")
    print("  ✓ Camadas concêntricas: dependência sempre para dentro")


def demo_hexagonal():
    print("\n" + "─" * 55)
    print("  3. HEXAGONAL (PORTS & ADAPTERS)")
    print("     Portas (interfaces) + Adaptadores (concretos)")
    print("─" * 55)

    repo = PedidoRepositoryEmMemoria()
    handler = HexagonalPedidoHandler(repo)

    cmd = {
        "pedido_id": 3001,
        "cliente": "Roberto Lima",
        "itens": [
            {"id": "P05", "nome": "SSD 1TB", "preco": 600.00, "qtd": 2},
            {"id": "P06", "nome": "Cabo HDMI", "preco": 45.00, "qtd": 3},
        ],
    }
    result = handler.handle_criar_pedido(cmd)
    print(f"  Comando executado: {result}")
    consulta = handler.handle_consultar_pedido(3001)
    print(f"  Consulta: {consulta}")
    print("  ✓ Domínio 100% isolado por portas; adaptadores trocáveis")


if __name__ == "__main__":
    print("=" * 60)
    print("  CLEAN × ONION × HEXAGONAL — MESMO DOMÍNIO")
    print("=" * 60)
    print("\n  Domínio comum: Pedido + ItemPedido + Dinheiro (Value Object)")

    demo_clean()
    demo_onion()
    demo_hexagonal()

    print("\n" + "=" * 60)
    print("  COMPARATIVO:")
    print("  ┌──────────┬──────────────────────────────────┐")
    print("  │ Clean    │ Entities → UseCases → Adapters   │")
    print("  │ Onion    │ Domain ← App ← Infra ← Present.  │")
    print("  │ Hexagonal│ Ports (interfaces) + Adapters     │")
    print("  ├──────────┼──────────────────────────────────┤")
    print("  │ TODOS    │ Domínio isolado de frameworks     │")
    print("  │ TODOS    │ Dependência aponta para dentro    │")
    print("  │ TODOS    │ Testabilidade e troca de infra    │")
    print("  └──────────┴──────────────────────────────────┘")
    print("=" * 60)
